sensors/baro/ms5611.c

```diff
@@ -0,0 +1,280 @@
+	uint32_t D1 = 0, D2 = 0;
+	int32_t dT;
+	int64_t off, sens;
+
+	/* Initialize temperature with too big delay for certain response */
+	while (ms5611_measure(ctx, (CMD_CVRT_TEMP | CVRT_OSR_4096), OSR_4096_SLEEP * 2, &D2) < 0)
+		;
+
+	/* Translate temperature */
+	dT = (int32_t)D2 - (((int32_t)ctx->prom[5]) << 8);           /* dT = D2 - C5 * 2^8 */
+	temp = 2000 + (((int64_t)dT * (int64_t)ctx->prom[6]) >> 23); /* TEMP = 2000 + dT * C6 / 2^23 */
+
+	while (1) {
+		/* Read pressure */
+		if (ms5611_measure(ctx, (CMD_CVRT_PRESS | CVRT_OSR_4096), OSR_4096_SLEEP, &D1) < 0) {
+			continue;
+		}
+
+		gettime(&now, NULL);
+
+		/* Read temperature. Does not repeat failed read - fall back to old temperature */
+		if (ms5611_measure(ctx, (CMD_CVRT_TEMP | CVRT_OSR_1024), OSR_1024_SLEEP, &D2) == 0) {
+			/* Translate temperature */
+			dT = (int32_t)D2 - (((int32_t)ctx->prom[5]) << 8);           /* dT = D2 - C5 * 2^8 */
+			temp = 2000 + (((int64_t)dT * (int64_t)ctx->prom[6]) >> 23); /* TEMP = 2000 + dT * C6 / 2^23 */
+		}
+
+		/* Translate press */
+		off = ((int64_t)ctx->prom[2] << 16) + (((int64_t)dT * (int64_t)ctx->prom[4]) >> 7);  /* OFF = C2 * 2^16 + (C4 * dT) / 2^7 */
+		sens = ((int64_t)ctx->prom[1] << 15) + (((int64_t)dT * (int64_t)ctx->prom[3]) >> 8); /* SENS = C1 * 2^15 + (C4 * dT) / 2^8 */
+		press = ((((int64_t)D1 * sens) >> 21) - off) >> 15;                                  /* P = ((D1 * SENS) / 2^21 - OFF) / 2^15 */
+
+		/* MS5611 operating ranges: temperature from -40C to +80C, and pressure: 45000Pa, 110000 Pa */
+		if (press > 45000 && press < 120000 && temp > -4000 && temp < 8500) {
+			ctx->evtBaro.timestamp = now;
+			ctx->evtBaro.baro.pressure = press;
+			ctx->evtBaro.baro.temp = (temp / 100) + 273;
+
+			sensors_publish(info->id, &ctx->evtBaro);
+		}
+	}
+
+	return;
+}
+
+
+static int ms5611_hwSetup(ms5611_ctx_t *ctx)
+{
+	uint8_t cmd, i;
+
+	/* Reset ms5611 sequence */
+	cmd = CMD_RESET;
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), NULL, 0, 0) < 0) {
+		fprintf(stderr, "ms5611: failed to reset device\n");
+		return -1;
+	}
+	usleep(RESET_SLEEP);
+
+	/* PROM reading */
+	for (i = 0; i < PROM_SIZE; i++) {
+		cmd = CMD_READ_PROM | (i << 1); /* LSB of prom address is always 0 */
+		if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &cmd, sizeof(cmd), &ctx->prom[i], sizeof(ctx->prom[i]), sizeof(cmd)) < 0) {
+			fprintf(stderr, "ms5611: failed read PROM C%i\n", i + 1);
+			return -1;
+		}
+		ctx->prom[i] = (ctx->prom[i] << 8) | (ctx->prom[i] >> 8); /* swap higher and lower bytes */
+	}
+
+	/* Check CRC of PROM (lower nibble of last prom byte) with calculated CRC */
+	if (ms5611_crc4(ctx->prom) != (ctx->prom[PROM_SIZE - 1] & 0x000f)) {
+		fprintf(stderr, "ms5611: wrong PROM crc\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int ms5611_start(sensor_info_t *info)
+{
+	int err;
+	ms5611_ctx_t *ctx = (ms5611_ctx_t *)info->ctx;
+
+	err = beginthread(ms5611_publishthr, 4, ctx->stack, sizeof(ctx->stack), info);
```

Comment:
Priority of 4 seems low to me

Reply:
All sensors currently work on priority 4. It does not impose observable overhead in short term.